packages/transcript-view/src/transcript-view.ts

```diff
@@ -0,0 +1,383 @@
+    const autoScrollChangedEvent = new CustomEvent('autoScrollChanged', {
+      detail: { autoScroll: this.autoScroll },
+      bubbles: true,
+      composed: true,
+    });
+    this.dispatchEvent(autoScrollChangedEvent);
+  }
+
+  private scrollToActiveEntry(): void {
+    if (!this.autoScroll) {
+      return;
+    }
+    const { activeTranscriptEntry } = this;
+    if (!activeTranscriptEntry) {
+      return;
+    }
+    this.scrollToElement(activeTranscriptEntry);
+  }
+
+  private scrollToSelectedSearchResult(): void {
+    const { selectedSearchResult } = this;
+    if (!selectedSearchResult) {
+      return;
+    }
+    this.autoScroll = false;
+    this.scrollToElement(selectedSearchResult);
+  }
+
+  private scrollToElement(element: HTMLElement): void {
+    const { scrollView } = this;
+    if (!scrollView) {
+      return;
+    }
+
+    const scrollContainerRect = scrollView.getBoundingClientRect();
+    const activeEntryRect = element.getBoundingClientRect();
+    const scrollContainerHeight = scrollContainerRect.height;
+    const { topContextHeight } = this;
+    const { bottomContextHeight } = this;
+    const focusBottom = scrollContainerHeight - bottomContextHeight;
+
+    // if the active entry is above the top context area or below the bottom of the focus area,
+    // scroll it to the top of the focus area
+    if (
+      activeEntryRect.bottom > scrollContainerRect.top + focusBottom ||
+      activeEntryRect.top < scrollContainerRect.top
+    ) {
+      // eslint-disable-next-line max-len
+      const newTargetScrollPos =
+        activeEntryRect.top - scrollContainerRect.top + scrollView.scrollTop - topContextHeight;
+      this.scrollToOffsetWithDuration(newTargetScrollPos, 1);
+    }
+  }
+
+  private updateTimePosition(): void {
+    const activeEntry = this.activeTranscriptEntry;
+    if (!activeEntry) {
+      return;
+    }
+
+    const parentNode = activeEntry.parentNode as HTMLElement;
+    const parentOffset = parentNode.getBoundingClientRect();
+    const offset = activeEntry.getBoundingClientRect().top - parentOffset.top;
+
+    this.timeScrollTop = offset;
+  }
+
+  private scrollToOffsetWithDuration(offset: number, duration: number, onDone?: () => void): void {
+    const { scrollView } = this;
+    if (!scrollView) {
+      return;
+    }
+
+    const start = scrollView.scrollTop;
+    const change = offset - start;
+    const startTime = performance.now();
+    let now;
+    let elapsed;
+    let t;
+
+    function easeInOutQuad(time: number): number {
+      return time < 0.5 ? 2 * time * time : -1 + (4 - 2 * time) * time;
+    }
+
+    function animateScroll(): void {
+      if (!scrollView) {
+        return;
+      }
+
+      now = performance.now();
+      elapsed = (now - startTime) / 1000;
+      t = elapsed / duration;
+
+      scrollView.scrollTop = start + change * easeInOutQuad(t);
+
+      if (t < 1) {
+        window.requestAnimationFrame(animateScroll);
+      } else if (onDone) {
```

Comment:
Since you have a default value for `onDone`, won't this alway be true when `t >= 1`?

Reply:
`onDone` is an optional. The syntax in the method signature isn't defining a default value, it's just how you denote the function signature.

ie `onDone?: () => void` is saying "give me a function (optionally) that takes no parameters and doesn't return anything.

You could do `onDone?: (result: string) => void` which is saying "pass me a function (optionally) that takes a single `string` parameter and returns nothing"

TypeScript 3.7 is introducing this feature called "chained optionals" which will allow you to do something like this `onDone?()` which checks for its existence first, then executes if it exists.